In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 

# Set the style of the plots
sns.set_style("darkgrid")

In [13]:
df = pd.read_csv('../datasets\coaster_db.csv')
# Jab bhi DataFrame print karun, saari columns full dikhani hain. Kuch bhi hide ya '...' wala short form mat karna.
pd.set_option('display.max_columns', None)


In [14]:
df.head()

coaster_name          Length             Speed  \
0                      Switchback Railway  600 ft (180 m)  6 mph (9.7 km/h)   
1                       Flip Flap Railway             NaN               NaN   
2  Switchback Railway (Euclid Beach Park)             NaN               NaN   
3            Loop the Loop (Coney Island)             NaN               NaN   
4            Loop the Loop (Young's Pier)             NaN               NaN   

                         Location   Status   Opening date   Type  \
0                    Coney Island  Removed  June 16, 1884   Wood   
1                   Sea Lion Park  Removed           1895   Wood   
2  Cleveland, Ohio, United States   Closed            NaN  Other   
3                           Other  Removed           1901  Steel   
4                           Other  Removed           1901  Steel   

             Manufacturer Height restriction        Model        Height  \
0  LaMarcus Adna Thompson                NaN  Lift Packed  50 ft (15 m)   
1            Lina Beecher                NaN          NaN           NaN   
2                     NaN                NaN          NaN           NaN   
3          Edwin Prescott                NaN          NaN           NaN   
4          Edwin Prescott                NaN          NaN           NaN   

   Inversions Lift/launch system Cost  \
0         NaN            gravity  NaN   
1         1.0                NaN  NaN   
2         NaN                NaN  NaN   
3         1.0                NaN  NaN   
4         1.0                NaN  NaN   

                                              Trains  \
0                                                NaN   
1  a single car. Riders are arranged 1 across in ...   
2                                                NaN   
3  a single car. Riders are arranged 2 across in ...   
4                                                NaN   

                Park section Duration              Capacity G-force  \
0  Coney Island Cyclone Site     1:00  1600 riders per hour     2.9   
1                        NaN      NaN                   NaN      12   
2                        NaN      NaN                   NaN     NaN   
3                        NaN      NaN                   NaN     NaN   
4                        NaN      NaN                   NaN     NaN   

                 Designer Max vertical angle          Drop Soft opening date  \
0  LaMarcus Adna Thompson                30°  43 ft (13 m)               NaN   
1            Lina Beecher                NaN           NaN               NaN   
2                     NaN                NaN           NaN               NaN   
3         Edward A. Green                NaN           NaN               NaN   
4         Edward A. Green                NaN           NaN               NaN   

  Fast Lane available            Replaced            Track layout  \
0                 NaN                 NaN  Gravity pulled coaster   
1                 NaN                 NaN                     NaN   
2                 NaN                 NaN                     NaN   
3                 NaN  Switchback Railway                     NaN   
4                 NaN                 NaN                     NaN   

  Fastrack available Soft opening date.1 Closing date Opened  Replaced by  \
0                NaN                 NaN          NaN    NaN          NaN   
1                NaN                 NaN         1902    NaN          NaN   
2                NaN                 NaN          NaN   1895          NaN   
3                NaN                 NaN         1910    NaN  Giant Racer   
4                NaN                 NaN         1912    NaN          NaN   

  Website Flash Pass Available Must transfer from wheelchair Theme  \
0     NaN                  NaN                           NaN   NaN   
1     NaN                  NaN                           NaN   NaN   
2     NaN                  NaN                           NaN   NaN   
3     NaN                  NaN              

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1087 entries, 0 to 1086
Data columns (total 56 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   coaster_name                   1087 non-null   object 
 1   Length                         953 non-null    object 
 2   Speed                          937 non-null    object 
 3   Location                       1087 non-null   object 
 4   Status                         874 non-null    object 
 5   Opening date                   837 non-null    object 
 6   Type                           1087 non-null   object 
 7   Manufacturer                   1028 non-null   object 
 8   Height restriction             831 non-null    object 
 9   Model                          744 non-null    object 
 10  Height                         965 non-null    object 
 11  Inversions                     932 non-null    float64
 12  Lift/launch system             795 non-null    o

In [16]:
df.describe(include='all')

coaster_name          Length             Speed Location  \
count               1087             953               937     1087   
unique               990             569               243      280   
top     Batman: The Ride  935 ft (285 m)  50 mph (80 km/h)    Other   
freq                   7              21                63      250   
mean                 NaN             NaN               NaN      NaN   
std                  NaN             NaN               NaN      NaN   
min                  NaN             NaN               NaN      NaN   
25%                  NaN             NaN               NaN      NaN   
50%                  NaN             NaN               NaN      NaN   
75%                  NaN             NaN               NaN      NaN   
max                  NaN             NaN               NaN      NaN   

           Status Opening date   Type Manufacturer Height restriction   Model  \
count         874          837   1087         1028                831     744   
unique         15          656     98          102                100     317   
top     Operating         1976  Steel       Vekoma     48 in (122 cm)  Custom   
freq          668            7    308          135                224      20   
mean          NaN          NaN    NaN          NaN                NaN     NaN   
std           NaN          NaN    NaN          NaN                NaN     NaN   
min           NaN          NaN    NaN          NaN                NaN     NaN   
25%           NaN          NaN    NaN          NaN                NaN     NaN   
50%           NaN          NaN    NaN          NaN                NaN     NaN   
75%           NaN          NaN    NaN          NaN                NaN     NaN   
max           NaN          NaN    NaN          NaN                NaN     NaN   

              Height  Inversions Lift/launch system         Cost  \
count            965  932.000000                795          382   
unique           382         NaN                116          272   
top     70 ft (21 m)         NaN    Chain lift hill  $10 million   
freq              20         NaN                411            6   
mean             NaN    1.547210                NaN          NaN   
std              NaN    2.114073                NaN          NaN   
min              NaN    0.000000                NaN          NaN   
25%              NaN    0.000000                NaN          NaN   
50%              NaN    0.000000                NaN          NaN   
75%              NaN    3.000000                NaN          NaN   
max              NaN   14.000000                NaN          NaN   

                                                   Trains   Park section  \
count                                                 718            487   
unique                                                221            271   
top     2 trains with 6 cars. Riders are arranged 2 ac...  Planet Snoopy   
freq                                                   52             14   
mean                                                  NaN            NaN   
std                                                   NaN            NaN   
min                                                   NaN            NaN   
25%                                                   NaN            NaN   
50%                                                   NaN            NaN   
75%                                                   NaN            NaN   
max                                                   NaN            NaN   

       Duration              Capacity G-force        Designer  \
count       765                   575     362             578   
unique      208                   160      70             153   
top        1:30  1200 riders per hour       4  Werner Stengel   
freq         59                    36      46             172   
mean        NaN                   NaN     NaN             NaN   
std         NaN                   NaN     NaN             NaN   
m

In [17]:
num_columns = df.select_dtypes(include=['int64', 'float64']).columns
cat_columns = df.select_dtypes(include=['object']).columns

In [18]:
for col in num_columns:
    df[col] = df[col].fillna(df[col].median())

for col in cat_columns:
    df[col] = df[col].fillna(df[col].mode()[0])

df.sample(50)

coaster_name  \
386         Viper (Six Flags Great Adventure)   
431                        Tower of Terror II   
137                             Texas Cyclone   
961                             Cobra's Curse   
108                    Toboggan (Hersheypark)   
688   Powder Keg: A Blast into the Wilderness   
737                 Mick Doohan's Motocoaster   
66        Thunderbolt (Six Flags New England)   
981              Dawson Duel (roller coaster)   
779                  Tsunami (roller coaster)   
794                  Manta (SeaWorld Orlando)   
1006                           Incredicoaster   
873                            Polar X-plorer   
96                                     Galaxi   
539                                 Mega Zeph   
384                       T2 (roller coaster)   
635                             Ozark Wildcat   
475             Exterminator (roller coaster)   
900                          Nitro (Imagicaa)   
606                            Primeval Whirl   
348                          Batman: The Ride   
444                       Journey to Atlantis   
630                Half Pipe (roller coaster)   
520                           Catwoman's Whip   
106         Rock n' Roller Coaster (Opryland)   
492            Nightmare at Crack Axle Canyon   
967            Lightning Rod (roller coaster)   
437             Blue Tornado (roller coaster)   
358                      Raptor (Cedar Point)   
22                    Jack Rabbit (Kennywood)   
351                      Corkscrew (Playland)   
205                       Alpenexpress Enzian   
1002                   Fēnix (roller coaster)   
581         Nitro (Six Flags Great Adventure)   
611                 Scooby-Doo Spooky Coaster   
8                                Drop the Dip   
280                           Olympia Looping   
176           American Eagle (roller coaster)   
457                  Scooby's Ghoster Coaster   
654                Lightning (roller coaster)   
534                   Kraken (roller coaster)   
506            Thunderbolt (Celebration City)   
1074                Defiance (roller coaster)   
259              Boomerang (Six Flags México)   
818              The Monster (Walygator Parc)   
151        Loch Ness Monster (roller coaster)   
878                The Swarm (roller coaster)   
894               GateKeeper (roller coaster)   
395           The Great White (Morey's Piers)   
703          Goliath (Six Flags Over Georgia)   

                                Length                   Speed  \
386                   1,670 ft (510 m)        48 mph (77 km/h)   
431                 376.4 m (1,235 ft)  160.9 km/h (100.0 mph)   
137                   3,180 ft (970 m)        50 mph (80 km/h)   
961                   2,100 ft (640 m)        40 mph (64 km/h)   
108                     450 ft (140 m)        50 mph (80 km/h)   
688                 3,506 ft (1,069 m)       64 mph (103 km/h)   
737                   605 m (1,985 ft)        72 km/h (45 mph)   
66                    2,600 ft (790 m)        40 mph (64 km/h)   
981               1,607.6 ft (490.0 m)    24.9 mph (40.1 km/h)   
779                     935 ft (285 m)        55 mph (89 km/h)   
794                 3,359 ft (1,024 m)        56 mph (90 km/h)   
1006                6,072 ft (1,851 m)        55 mph (89 km/h)   
873                     935 ft (285 m)        65 km/h (40 mph)   
96                      935 ft (285 m)        50 mph (80 km/h)   
539                 4,000 ft (1,200 m)        57 mph (92 km/h)   
384               2,171.9 ft (662.0 m)    49.7 mph (80.0 km/h)   
635                   2,600 ft (790 m)        45 mph (72 km/h)   
475               1,377.9 ft (420.0 m)    29.1 mph (46.8 km/h)   
900                   2,800 ft (850 m)   65.2 mph (104.9 km/h)   
606               1,377.9 ft (420.0 m)    29.1 mph (46.8 km/h)   
348   2,693 or 2,700 ft (821 or 823 m)        50 mph (80 km/h)   
444                     935 ft (285 m)        50 mph (80 km/h)   
630            

In [19]:
df.drop_duplicates(inplace=True)
df.sample(50)

coaster_name                    Length  \
169                 Corkscrew (Alton Towers)          2,460 ft (750 m)   
307                           Turbo Mountain          420 m (1,380 ft)   
292                           Thunder Looper            935 ft (285 m)   
393                Fujiyama (roller coaster)        6,709 ft (2,045 m)   
491                                 Montezum        1,030 m (3,380 ft)   
362                      Titan (Space World)        5,019 ft (1,530 m)   
273                   Excalibur (Valleyfair)          2,415 ft (736 m)   
769               Possessed (roller coaster)            704 ft (215 m)   
368                         The Gold Coaster          900 m (3,000 ft)   
909       Woodstock Express (Kings Dominion)          1,385 ft (422 m)   
444                      Journey to Atlantis            935 ft (285 m)   
590            Wildfire (Silver Dollar City)          3,073 ft (937 m)   
89                      Dahlonega Mine Train          2,327 ft (709 m)   
1058          Dragon Slayer (roller coaster)            770 ft (230 m)   
171                         Judge Roy Scream          2,670 ft (810 m)   
830           Black Diamond (roller coaster)            935 ft (285 m)   
418   The Great White (SeaWorld San Antonio)          2,562 ft (781 m)   
296                Ultra Twister (Six Flags)          1,421 ft (433 m)   
48                      Zip (roller coaster)          2,500 ft (760 m)   
463                                Vogel Rok          2,296 ft (700 m)   
93    Mini Mine Train (Six Flags Over Texas)            935 ft (285 m)   
686                  American Dreier Looping        3,444 ft (1,050 m)   
848                             Sky Scrapper      2,805.1 ft (855.0 m)   
1004               Hyperion (roller coaster)        1,450 m (4,760 ft)   
36                              Jazz Railway          1,200 ft (370 m)   
272                 Eurosat - CanCan Coaster          900 m (3,000 ft)   
350                  Condor (roller coaster)          662 m (2,172 ft)   
119                                 Racer 75    3,368.5 ft (1,026.7 m)   
737                Mick Doohan's Motocoaster          605 m (1,985 ft)   
699               Eejanaika (roller coaster)  3,782.83 ft (1,153.01 m)   
719                                X-Coaster            492 ft (150 m)   
797              Roller Coaster (Papéa Parc)      1,181.1 ft (360.0 m)   
385                               Twister II        4,640 ft (1,410 m)   
477          Great Chase (Six Flags America)             280 ft (85 m)   
553                        Steel Dragon 2000        2,479 m (8,133 ft)   
777                              Steel Lasso            293 m (961 ft)   
619                           Wicked Twister            675 ft (206 m)   
482                               Iron Gwazi        4,075 ft (1,242 m)   
865               Dragonfly (roller coaster)          361 m (1,184 ft)   
95                             Daidarasaurus        2,340 m (7,680 ft)   
381              Sidewinder (Elitch Gardens)            635 ft (194 m)   
906                            Storm Coaster          470 m (1,540 ft)   
557                      Svalbard Ekspressen        679.2 ft (207.0 m)   
207                              Astro Storm          1,490 ft (450 m)   
239                                La Vibora          1,490 ft (450 m)   
634                          Nemesis Inferno          750 m (2,460 ft)   
852                              Timber Drop            935 ft (285 m)   
164         Rolling Thunder (roller coaster)        3,200 ft (975.4 m)   
320     Flashback (Six Flags Magic Mountain)          1,900 ft (580 m)   
377        Flight Deck (Canada's Wonderland)     689.01 m (2,260.5 ft)   

                       Speed                         Location  \
169         40 mph (64 km/h)                     Alton Towers   
307         57 km/h (35 mph)                  Adventure World   
292         50 mph (80 km/h)                            Other 

In [20]:
"""
for col in cat_columns:

    # convert values to string ........
    df[col] = df[col].astype(str)

    # remove extra spaces from start and end .........
    df[col] = df[col].str.strip()

    # convert to lowercase ........
    df[col] = df[col].str.lower()
"""

for col in cat_columns:
    df[col] = df[col].astype(str).str.strip().str.lower()

df.sample(20)

coaster_name                  Length  \
926                    ragin' cajun (roller coaster)    1,377.9 ft (420.0 m)   
476                                 georgia scorcher        3,000 ft (910 m)   
69                little dipper (conneaut lake park)          935 ft (285 m)   
208                    big bad wolf (roller coaster)        2,800 ft (850 m)   
867                  goliath (six flags new england)    1,204.1 ft (367.0 m)   
114   runaway mine train (six flags great adventure)          935 ft (285 m)   
261                                       crazy bats  3,851.7 ft (1,174.0 m)   
1002                          fēnix (roller coaster)        2,667 ft (813 m)   
45                          lightning (revere beach)          935 ft (285 m)   
2             switchback railway (euclid beach park)          935 ft (285 m)   
273                           excalibur (valleyfair)        2,415 ft (736 m)   
370              cannonball express (roller coaster)        1,738 ft (530 m)   
669                       time warp (roller coaster)   390.99 m (1,282.8 ft)   
710                          patriot (worlds of fun)        3,081 ft (939 m)   
913              the bat (kings island; opened 1993)        2,352 ft (717 m)   
271                        anaconda (walygator parc)      3,937 ft (1,200 m)   
1020               wonder woman golden lasso coaster        1,800 ft (550 m)   
342                  lightning bolt (roller coaster)        2,400 ft (730 m)   
990                                   mystic timbers        3,265 ft (995 m)   
993                       red force (roller coaster)        880 m (2,890 ft)   

                      Speed                            Location     Status  \
926    29.1 mph (46.8 km/h)                               other  operating   
476        54 mph (87 km/h)              six flags over georgia  operating   
69         50 mph (80 km/h)                  conneaut lake park  operating   
208        48 mph (77 km/h)          busch gardens williamsburg    removed   
867   65.6 mph (105.6 km/h)                               other  operating   
114        50 mph (80 km/h)  jackson, new jersey, united states  operating   
261    28.9 mph (46.5 km/h)                       phantasialand  operating   
1002       59 mph (95 km/h)                           toverland  operating   
45         50 mph (80 km/h)                        revere beach    removed   
2          50 mph (80 km/h)      cleveland, ohio, united states     closed   
273    54.5 mph (87.7 km/h)                          valleyfair  operating   
370        34 mph (55 km/h)                  pleasurewood hills  operating   
669    41.4 km/h (25.7 mph)                 canada's wonderland  operating   
710        60 mph (97 km/h)                       worlds of fun  operating   
913        51 mph (82 km/h)                        kings island  operating   
271        50 mph (80 km/h)                      walygator parc  operating   
1020       52 mph (84 km/h)              six flags fiesta texas  operating   
342        35 mph (56 km/h)     mgm grand adventures theme park    removed   
990        53 mph (85 km/h)                        kings island  operating   
993      180 km/h (112 mph)                        ferrari land  operating   

                     Opening date                           Type  \
926                          1972  steel – wild mouse – spinning   
476                   may 8, 1999               steel – stand-up   
69                          1950s                          steel   
208                 june 15, 1984              steel – suspended   
867                          1972     steel – shuttle – inverted   
114                          1972                          other   
261               april 1st, 1988               steel – enclosed   
1002                  7 july 2018           steel – wing coaster   
45                   july 2, 1927                           wood   
2                            1972       

In [21]:
df.to_csv('../datasets/cleaned_data.csv', index=False)

In [22]:
df.head()

coaster_name          Length             Speed  \
0                      switchback railway  600 ft (180 m)  6 mph (9.7 km/h)   
1                       flip flap railway  935 ft (285 m)  50 mph (80 km/h)   
2  switchback railway (euclid beach park)  935 ft (285 m)  50 mph (80 km/h)   
3            loop the loop (coney island)  935 ft (285 m)  50 mph (80 km/h)   
4            loop the loop (young's pier)  935 ft (285 m)  50 mph (80 km/h)   

                         Location   Status   Opening date   Type  \
0                    coney island  removed  june 16, 1884   wood   
1                   sea lion park  removed           1895   wood   
2  cleveland, ohio, united states   closed           1972  other   
3                           other  removed           1901  steel   
4                           other  removed           1901  steel   

             Manufacturer Height restriction        Model        Height  \
0  lamarcus adna thompson     48 in (122 cm)  lift packed  50 ft (15 m)   
1            lina beecher     48 in (122 cm)    boomerang  70 ft (21 m)   
2                  vekoma     48 in (122 cm)    boomerang  70 ft (21 m)   
3          edwin prescott     48 in (122 cm)    boomerang  70 ft (21 m)   
4          edwin prescott     48 in (122 cm)    boomerang  70 ft (21 m)   

   Inversions Lift/launch system         Cost  \
0         0.0            gravity  $10 million   
1         1.0    chain lift hill  $10 million   
2         0.0    chain lift hill  $10 million   
3         1.0    chain lift hill  $10 million   
4         1.0    chain lift hill  $10 million   

                                              Trains  \
0  2 trains with 6 cars. riders are arranged 2 ac...   
1  a single car. riders are arranged 1 across in ...   
2  2 trains with 6 cars. riders are arranged 2 ac...   
3  a single car. riders are arranged 2 across in ...   
4  2 trains with 6 cars. riders are arranged 2 ac...   

                Park section Duration              Capacity G-force  \
0  coney island cyclone site     1:00  1600 riders per hour     2.9   
1              planet snoopy     1:30  1200 riders per hour      12   
2              planet snoopy     1:30  1200 riders per hour       4   
3              planet snoopy     1:30  1200 riders per hour       4   
4              planet snoopy     1:30  1200 riders per hour       4   

                 Designer Max vertical angle           Drop Soft opening date  \
0  lamarcus adna thompson                30°   43 ft (13 m)    april 29, 2005   
1            lina beecher                90°  100 ft (30 m)    april 29, 2005   
2          werner stengel                90°  100 ft (30 m)    april 29, 2005   
3         edward a. green                90°  100 ft (30 m)    april 29, 2005   
4         edward a. green                90°  100 ft (30 m)    april 29, 2005   

   Fast Lane available                                Replaced  \
0  fast lane available  mine train through nature's wonderland   
1  fast lane available  mine train through nature's wonderland   
2  fast lane available  mine train through nature's wonderland   
3  fast lane available                      switchback railway   
4  fast lane available  mine train through nature's wonderland   

             Track layout  Fastrack available Soft opening date.1  \
0  gravity pulled coaster  fastrack available      april 29, 2005   
1                 twister  fastrack available      april 29, 2005   
2                 twister  fastrack available      april 29, 2005   
3                 twister  fastrack available      april 29, 2005   
4                 twister  fastrack available      april 29, 2005   

         Closing date Opened  Replaced by           Website  \
0  september 16, 2007   1895    the joker  official website   
1                1902   1895    the joker  official website   
2  september 16, 2007   1895    the joker  official website   
3                1910   1895  giant racer  official website   
4                1912

In [25]:
# print(df.quantile([0.25, 0.5, 0.75]))

steps to Find Outliers Using IQR [interquartile range]:

Find Q1 and Q3

→ Use data.quantile(0.25) and data.quantile(0.75)

Calculate IQR

→ IQR = Q3 - Q1

Set boundaries:

Lower Bound = Q1 – 1.5 × IQR

Upper Bound = Q3 + 1.5 × IQR

Any value below the lower bound OR above the upper bound is considered an OUTLIER.

 The 1.5 multiplier is a standard rule (from statistician John Tukey). You can use 3.0 for "extreme outliers", but 1.5 is common.